pkg/server/instancecleaner/cleaner.go

```diff
@@ -0,0 +1,91 @@
+package instancecleaner
+
+import (
+	"context"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/gocardless/draupnir/pkg/exec"
+	"github.com/gocardless/draupnir/pkg/models"
+	"github.com/gocardless/draupnir/pkg/store"
+	"github.com/prometheus/common/log"
+
+	google "google.golang.org/api/oauth2/v1"
+)
+
+type InstanceCleaner struct {
+	done          chan bool
+	logger        log.Logger
+	instanceStore store.InstanceStore
+	executor      exec.Executor
+}
+
+func NewInstanceCleaner(logger log.Logger, instanceStore store.InstanceStore, executor exec.Executor) *InstanceCleaner {
+	done := make(chan bool)
+	return &InstanceCleaner{
+		done:          done,
+		logger:        logger.With("component", "cleaner"),
+		instanceStore: instanceStore,
+		executor:      executor,
+	}
+}
+
+func (ic *InstanceCleaner) Start(cctx context.Context, interval time.Duration) error {
+	for {
+		select {
+		case <-time.After(interval):
+			ic.logger.Info("cleaning old instances with invalid tokens")
+			instances, err := ic.instanceStore.List()
+			if err != nil {
+				ic.logger.Info("skipping cleaning old instances - could not get instances list")
+			} else {
+				for _, instance := range instances {
+					if instance.AccessToken != "" {
+						valid, err := validateAccessToken(instance.AccessToken)
+						if err != nil {
+							ic.logger.With("instance", instance.ID).Error("failed to validate token: %s", err.Error())
+						} else if !valid {
+							ic.logger.With("instance", instance.ID).Info("token for instance invalid - destroying instance")
+							err = ic.destroyInstance(instance)
+							if err != nil {
+								ic.logger.With("instance", instance.ID).Error("failed to destroy instance: %s", err.Error())
+							}
+						}
+					}
+				}
+			}
+		case <-cctx.Done():
+			ic.done <- true
+			return nil
+		}
+	}
+}
+
+func (ic *InstanceCleaner) Stop() {
+	<-ic.done
+}
+
+func (ic *InstanceCleaner) destroyInstance(instance models.Instance) error {
+	err := ic.executor.DestroyInstance(context.Background(), instance.ID)
+	if err == nil {
+		err = ic.instanceStore.Destroy(instance)
+	}
+	return err
+}
+
+func validateAccessToken(accessToken string) (bool, error) {
+	service, err := google.New(http.DefaultClient)
```

Comment:
You'd probably want to make the google client a field on the `InstanceCleaner` struct, so you can stub it with a client you control in tests. Given you have an initialiser (`NewInstanceCleaner`) there's no harm done to leaving it.